src/constraint/CollisionFree.cpp

```diff
@@ -27,35 +27,74 @@ statespace::StateSpacePtr CollisionFree::getStateSpace() const
 
 //==============================================================================
 bool CollisionFree::isSatisfied(
-    const aikido::statespace::StateSpace::State* _state) const
+    const aikido::statespace::StateSpace::State* _state,
+    TestableOutcome* outcome) const
 {
+  CollisionFreeOutcome* collisionFreeOutcome = nullptr;
+  if (outcome)
+  {
+    collisionFreeOutcome = dynamic_cast<CollisionFreeOutcome*>(outcome);
+    if (!collisionFreeOutcome)
+      throw std::invalid_argument(
+          "TestableOutcome pointer is not of type CollisionFreeOutcome.");
+    // Clear the outcome object in case it's being reused.
+    collisionFreeOutcome->clear();
+  }
+
   auto skelStatePtr = static_cast<const aikido::statespace::dart::
                                       MetaSkeletonStateSpace::State*>(_state);
   mStatespace->setState(skelStatePtr);
 
   bool collision = false;
   dart::collision::CollisionResult collisionResult;
-  for (auto groups : mGroupsToPairwiseCheck)
+  for (const auto& groups : mGroupsToPairwiseCheck)
   {
     collision = mCollisionDetector->collide(
         groups.first.get(),
         groups.second.get(),
         mCollisionOptions,
         &collisionResult);
+
     if (collision)
+    {
+      if (collisionFreeOutcome)
+      {
+        const auto& contacts = collisionResult.getContacts();
+        for (const auto& elem : contacts)
+        {
+          collisionFreeOutcome->markPairwiseContact(elem);
+        }
+      }
       return false;
+    }
   }
 
-  for (auto group : mGroupsToSelfCheck)
+  for (const auto& group : mGroupsToSelfCheck)
   {
     collision = mCollisionDetector->collide(
         group.get(), mCollisionOptions, &collisionResult);
     if (collision)
+    {
+      if (collisionFreeOutcome)
+      {
+        const auto& contacts = collisionResult.getContacts();
+        for (const auto& elem : contacts)
+        {
+          collisionFreeOutcome->markSelfContact(elem);
+        }
+      }
```

Comment:
Nit: If you make `CollisionFreeOutcome` a `friend class`, then this could be reduced to:
```c++
collisionFreeOutcome->mSelfContacts = collisionResult.getContacts();
```